Compilation fails

Hi,

I'm trying to compile your very interesting software (given the papers) on an Ubuntu 16.04 system.

It fails, as follows. Any ideas?

`-- The C compiler identification is GNU 7.2.0
-- The CXX compiler identification is GNU 7.2.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No build type selected, default to Release
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
-- Build files have been written to: src/c/ARTSynchronized/build

$ make
Scanning dependencies of target ARTSynchronized
[ 16%] Building CXX object CMakeFiles/ARTSynchronized.dir/OptimisticLockCoupling/Tree.cpp.o
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp: In member function ‘bool ART_OLC::Tree::lookupRange(const Key&, const Key&, Key&, TID*, std::size_t, std::size_t&, ART::ThreadInfo&) const’:
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:91:14: error: ‘function’ is not a member of ‘std’
         std::function<void(const N *)> copy = [&result, &resultSize, &resultsFound, &toContinue, &copy](const N *node) {
              ^~~~~~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:91:14: note: suggested alternative: ‘is_function’
         std::function<void(const N *)> copy = [&result, &resultSize, &resultsFound, &toContinue, &copy](const N *node) {
              ^~~~~~~~
              is_function
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:91:23: error: expected primary-expression before ‘void’
         std::function<void(const N *)> copy = [&result, &resultSize, &resultsFound, &toContinue, &copy](const N *node) {
                       ^~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:112:14: error: ‘function’ is not a member of ‘std’
         std::function<void(N *, uint8_t, uint32_t, const N *, uint64_t)> findStart = [&copy, &start, &findStart, &toContinue, this](
              ^~~~~~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:112:14: note: suggested alternative: ‘is_function’
         std::function<void(N *, uint8_t, uint32_t, const N *, uint64_t)> findStart = [&copy, &start, &findStart, &toContinue, this](
              ^~~~~~~~
              is_function
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:112:71: error: expression list treated as compound expression in functional cast [-fpermissive]
         std::function<void(N *, uint8_t, uint32_t, const N *, uint64_t)> findStart = [&copy, &start, &findStart, &toContinue, this](
                                                                       ^
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:112:23: error: expected primary-expression before ‘void’
         std::function<void(N *, uint8_t, uint32_t, const N *, uint64_t)> findStart = [&copy, &start, &findStart, &toContinue, this](
                       ^~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:181:14: error: ‘function’ is not a member of ‘std’
         std::function<void(N *, uint8_t, uint32_t, const N *, uint64_t)> findEnd = [&copy, &end, &toContinue, &findEnd, this](
              ^~~~~~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:181:14: note: suggested alternative: ‘is_function’
         std::function<void(N *, uint8_t, uint32_t, const N *, uint64_t)> findEnd = [&copy, &end, &toContinue, &findEnd, this](
              ^~~~~~~~
              is_function
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:181:71: error: expression list treated as compound expression in functional cast [-fpermissive]
         std::function<void(N *, uint8_t, uint32_t, const N *, uint64_t)> findEnd = [&copy, &end, &toContinue, &findEnd, this](
                                                                       ^
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:181:23: error: expected primary-expression before ‘void’
         std::function<void(N *, uint8_t, uint32_t, const N *, uint64_t)> findEnd = [&copy, &end, &toContinue, &findEnd, this](
                       ^~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:280:21: error: ‘copy’ was not declared in this scope
                     copy(node);
                     ^~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:280:21: note: suggested alternative:
In file included from /usr/include/c++/7/bits/locale_facets.h:48:0,
                 from /usr/include/c++/7/bits/basic_ios.h:37,
                 from /usr/include/c++/7/ios:44,
                 from /usr/include/c++/7/ostream:38,
                 from /usr/include/c++/7/iterator:64,
                 from /usr/include/tbb/concurrent_vector.h:42,
                 from /usr/include/tbb/enumerable_thread_specific.h:25,
                 from src/c/ARTSynchronized/OptimisticLockCoupling/../Epoche.h:6,
                 from src/c/ARTSynchronized/OptimisticLockCoupling/N.h:13,
                 from src/c/ARTSynchronized/OptimisticLockCoupling/Tree.h:7,
                 from src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:3:
/usr/include/c++/7/bits/streambuf_iterator.h:293:5: note:   ‘std::copy’
     copy(istreambuf_iterator<_CharT> __first,
     ^~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:294:33: error: ‘findStart’ was not declared in this scope
                                 findStart(n, k, level + 1, node, v);
                                 ^~~~~~~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:296:33: error: ‘copy’ was not declared in this scope
                                 copy(n);
                                 ^~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:296:33: note: suggested alternative:
In file included from /usr/include/c++/7/bits/locale_facets.h:48:0,
                 from /usr/include/c++/7/bits/basic_ios.h:37,
                 from /usr/include/c++/7/ios:44,
                 from /usr/include/c++/7/ostream:38,
                 from /usr/include/c++/7/iterator:64,
                 from /usr/include/tbb/concurrent_vector.h:42,
                 from /usr/include/tbb/enumerable_thread_specific.h:25,
                 from src/c/ARTSynchronized/OptimisticLockCoupling/../Epoche.h:6,
                 from src/c/ARTSynchronized/OptimisticLockCoupling/N.h:13,
                 from src/c/ARTSynchronized/OptimisticLockCoupling/Tree.h:7,
                 from src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:3:
/usr/include/c++/7/bits/streambuf_iterator.h:293:5: note:   ‘std::copy’
     copy(istreambuf_iterator<_CharT> __first,
     ^~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:298:33: error: ‘findEnd’ was not declared in this scope
                                 findEnd(n, k, level + 1, node, v);
                                 ^~~~~~~
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:79:91: warning: unused parameter ‘result’ [-Wunused-parameter]
     bool Tree::lookupRange(const Key &start, const Key &end, Key &continueKey, TID result[],
                                                                                           ^
src/c/ARTSynchronized/OptimisticLockCoupling/Tree.cpp:80:45: warning: unused parameter ‘resultSize’ [-Wunused-parameter]
                                 std::size_t resultSize, std::size_t &resultsFound, ThreadInfo &threadEpocheInfo) const {
                                             ^~~~~~~~~~
CMakeFiles/ARTSynchronized.dir/build.make:62: recipe for target 'CMakeFiles/ARTSynchronized.dir/OptimisticLockCoupling/Tree.cpp.o' failed
make[2]: *** [CMakeFiles/ARTSynchronized.dir/OptimisticLockCoupling/Tree.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/ARTSynchronized.dir/all' failed
make[1]: *** [CMakeFiles/ARTSynchronized.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
`

--- a/OptimisticLockCoupling/Tree.cpp
+++ b/OptimisticLockCoupling/Tree.cpp
@@ -1,5 +1,6 @@
 #include <assert.h>
 #include <algorithm>
+#include <functional>
 #include "Tree.h"
 #include "N.cpp"
 #include "../Epoche.cpp"

--- a/ROWEX/Tree.cpp
+++ b/ROWEX/Tree.cpp
@@ -1,5 +1,6 @@
 #include <assert.h>
 #include <algorithm>
+#include <functional>
 #include "Tree.h"
 #include "N.cpp"
 #include "../Epoche.cpp"

